### ISSUE TITLE
Flutter Web sourcemaps support

### ISSUE THREAD

**Issue by marandaneto** [getsentry/sentry-dart#417]
Related to:

https://github.com/flutter/flutter/issues/16883 
https://github.com/flutter/flutter/issues/37875

**Comment by ueman** [comment]
Dart has some extensions for source maps: https://github.com/dart-lang/sdk/blob/master/pkg/compiler/doc/sourcemap_extensions.md

@marandaneto Does this need to be considered in Sentry service?

**Comment by marandaneto** [comment]
@ueman probably, if we don't support such optimizations, we'd need to figure it out, I'll double-check that, thanks for the heads up.

**Comment by ueman** [comment]
For further reference, Dart's own implementation for getting readable stacktraces out Dart2JS compiled apps:
https://pub.dev/packages/source_map_stack_trace

**Comment by marandaneto** [comment]
although we can still improve this, it works well at some degree:

![Screenshot 2021-08-12 at 11 13 39](https://user-images.githubusercontent.com/5731772/129171939-3d38f613-3fea-46d0-9688-0d2d5584a3b5.png)

**Comment by venkatd** [comment]
@marandaneto are there any resources to troubleshoot why we can't see the lines of code?

This is what I am seeing in our app:
![image](https://user-images.githubusercontent.com/220686/130146638-b8c0ea53-efad-43c1-9c55-9108061f19e6.png)

This is just a link to our index.html file

Here's a subset of the logs on our GitHub actions:

<details>
<summary>logs</summary>

```
Run getsentry/action-release@v1.1.5
  with:
    environment: prod
    version: 9c223c34aa7b6b11e55b9ac06abffc021e33855b
    sourcemaps: /home/runner/work/***-frontend/***-frontend/packages/***_app/build/web
    finalize: true
  env:
    FLUTTER_ROOT: /opt/hostedtoolcache/flutter/b81165d259b4428af4efa47183842b577bd389c1/x64
    PUB_CACHE: /opt/hostedtoolcache/flutter/b81165d259b4428af4efa47183842b577bd389c1/x64/.pub-cache
    SENTRY_AUTH_TOKEN: ***
    SENTRY_ORG: ***
    SENTRY_PROJECT: ***
/usr/bin/docker run --name sentryintegrationssentrygithubactionreleaselatest_b4c352 --label c9e036 --workdir /github/workspace --rm -e FLUTTER_ROOT -e PUB_CACHE -e SENTRY_AUTH_TOKEN -e SENTRY_ORG -e SENTRY_PROJECT -e INPUT_ENVIRONMENT -e INPUT_VERSION -e INPUT_SOURCEMAPS -e INPUT_FINALIZE -e INPUT_STARTED_AT -e INPUT_VERSION_PREFIX -e INPUT_SET_COMMITS -e INPUT_PROJECTS -e INPUT_URL_PREFIX -e HOME -e GITHUB_JOB -e GITHUB_REF -e GITHUB_SHA -e GITHUB_REPOSITORY -e GITHUB_REPOSITORY_OWNER -e GITHUB_RUN_ID -e GITHUB_RUN_NUMBER -e GITHUB_RETENTION_DAYS -e GITHUB_ACTOR -e GITHUB_WORKFLOW -e GITHUB_HEAD_REF -e GITHUB_BASE_REF -e GITHUB_EVENT_NAME -e GITHUB_SERVER_URL -e GITHUB_API_URL -e GITHUB_GRAPHQL_URL -e GITHUB_WORKSPACE -e GITHUB_ACTION -e GITHUB_EVENT_PATH -e GITHUB_ACTION_REPOSITORY -e GITHUB_ACTION_REF -e GITHUB_PATH -e GITHUB_ENV -e RUNNER_DEBUG -e RUNNER_OS -e RUNNER_TOOL_CACHE -e RUNNER_TEMP -e RUNNER_WORKSPACE -e ACTIONS_RUNTIME_URL -e ACTIONS_RUNTIME_TOKEN -e ACTIONS_CACHE_URL -e GITHUB_ACTIONS=true -e CI=true -v "/var/run/docker.sock":"/var/run/docker.sock" -v "/home/runner/work/_temp/_github_home":"/github/home" -v "/home/runner/work/_temp/_github_workflow":"/github/workflow" -v "/home/runner/work/_temp/_runner_file_commands":"/github/file_commands" -v "/home/runner/work/***-frontend/***-frontend":"/github/workspace" sentryintegrations/sentry-github-action-release:latest
##[debug]Version is 9c223c34aa7b6b11e55b9ac06abffc021e33855b
##[debug]Setting commits with option 'auto'
##[debug]Adding sourcemaps
> Found 5 release files
> Analyzing 5 sources
> Rewriting sources
> Adding source map references
> Bundled 5 files for upload
> Uploaded release files to Sentry
> File upload complete (processing pending on server)

Source Map Upload Report
  Scripts
    /flutter_service_worker.js
    /main-9c223c34aa7b6b11e55b9ac06abffc021e33855b.dart.js
    /post-message.js
    /service-worker.js
  Source Maps
    /main-9c223c34aa7b6b11e55b9ac06abffc021e33855b.dart.js.map
```
</details>

**Comment by marandaneto** [comment]
@venkatd I'd open a new issue though, with a bit more context eg how (params) you've uploaded the source maps, if versioning and dist matches (the event itself and the source maps).

btw you can also try this: https://github.com/getsentry/sentry-dart-plugin

**Comment by venkatd** [comment]
@marandaneto thanks I just created an issue at #562 with logs of everything

We are using a GitHub action, but I feel like it's worth getting this working directly on `sentry-cli` as a first goal. It appears that the github action and the `sentry-dart-plugin` are both an extra layer on top of that.

Can move our discussion over to #562

**Comment by devxpy** [comment]
> although we can still improve this, it works well at some degree:
> 
> ![Screenshot 2021-08-12 at 11 13 39](https://user-images.githubusercontent.com/5731772/129171939-3d38f613-3fea-46d0-9688-0d2d5584a3b5.png)


Hey @marandaneto!

I'm not sure how you got to this point, can you please share the exact steps and code?

I followed these steps and got a pretty useless stacktrace - 
1. Created a new project at sentry and selected flutter
2. Downloaded https://github.com/getsentry/sentry-dart/tree/main/flutter/example
3. Added my sentry DSN to https://github.com/getsentry/sentry-dart/blob/fce129d973473dc8b7707b0a7da730925f23c3b8/flutter/example/lib/main.dart#L17
4. Added `org` and `project` to https://github.com/getsentry/sentry-dart/blob/fce129d973473dc8b7707b0a7da730925f23c3b8/flutter/example/pubspec.yaml#L43
5. Ran `flutter build web --release`
6. Ran `flutter packages pub run sentry_dart_plugin`
7. Removed this line because it wasn't allowing me to run the app https://github.com/getsentry/sentry-dart/blob/fce129d973473dc8b7707b0a7da730925f23c3b8/flutter/example/web/index.html#L19
8. Opened the app in Chrome (using `python3 http.server` in `build/web`)
9. Pressed the "Dart: try catch" button

<image src="https://user-images.githubusercontent.com/19492893/152437646-eebe3056-e7e6-4196-985e-122de61c4c55.png" height=800 />

**Comment by marandaneto** [comment]
@devxpy you've to pass `--source-maps` to your build command, https://github.com/getsentry/sentry-dart/blob/main/.github/workflows/web-example-ghpages.yml#L20

**Comment by marandaneto** [comment]
I've added to the README as well https://github.com/getsentry/sentry-dart-plugin/commit/14a37814b160e9ea819253306362f579775f7947#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5

**Comment by devxpy** [comment]
> I've added to the README as well [getsentry/sentry-dart-plugin@14a3781#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5](https://github.com/getsentry/sentry-dart-plugin/commit/14a37814b160e9ea819253306362f579775f7947#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5)

Thank you so much for the prompt reply :)

I got the source code to show up on sentry now!

Any ideas on how I can get it to show the correct exception type too? (It shows `minified:kP`)

<details>
<summary>Output</summary>

```
$ flutter build web --release --source-maps
Warning: You are using these overridden dependencies:
! sentry 6.3.0 from path ../../dart
! sentry_dio 6.3.0 from path ../../dio
! sentry_flutter 6.3.0 from path ..
Running "flutter pub get" in example...                          2,600ms

💪 Building with sound null safety 💪

Compiling lib/main.dart for the Web...                             30.9s
```

```
$ flutter packages pub run sentry_dart_plugin
☑ reading config values                                                             
☑ validating config values                                                             
[❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚] 18% uploading debug symbols..                                
includeNativeSources is disabled, not uploading sources.

stdout: > Found 0 debug information files
> No debug information files found

☑ uploading debug symbols                                                             
[❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚❚] 27% uploading source maps..                               
stdout: Created release sentry_flutter_example@6.3.0.


stdout: > Found 5 release files
> Analyzing 5 sources
> Rewriting sources
> Adding source map references
> Bundled 5 files for upload
> Uploaded release files to Sentry
> File processing complete
> Organization: dara-network
> Project: dara-app
> Release: sentry_flutter_example@6.3.0
> Dist: None

Source Map Upload Report
  Scripts
    ~/flutter_service_worker.js
  Minified Scripts
    ~/canvaskit/canvaskit.js (sourcemap at ../main.dart.js.map)
    ~/canvaskit/profiling/canvaskit.js (sourcemap at ../../main.dart.js.map)
    ~/main.dart.js (sourcemap at main.dart.js.map)
  Source Maps
    ~/main.dart.js.map


stdout: > Found 3 release files
> Analyzing 3 sources
> Rewriting sources
> Adding source map references
> Bundled 3 files for upload
> Uploaded release files to Sentry
> File processing complete
> Organization: dara-network
> Project: dara-app
> Release: sentry_flutter_example@6.3.0
> Dist: None

Source Map Upload Report
  Scripts
    ~/lib/generated_plugin_registrant.dart
    ~/lib/main.dart
    ~/lib/user_feedback_dialog.dart


stdout: Finalized release sentry_flutter_example@6.3.0.

☑ uploading source maps                                              
```

</details>



<image height=800 src="https://user-images.githubusercontent.com/19492893/152572688-754679da-55ea-429f-9187-2d36a7ccd5c5.png" />

**Comment by marandaneto** [comment]
That's a limitation right now, See https://docs.sentry.io/platforms/flutter/troubleshooting/#known-limitations
Look at the Obfuscate Caveat.

**Comment by devxpy** [comment]
Ah, dang! Thanks for linking the above. I guess your example was a debug build without obfuscation?

**Comment by a-eniou-pvotal** [comment]
any updates

**Comment by marandaneto** [comment]
So Web source maps should work already if you upload source maps, we're interested in collecting feedback in case there's any edge case that Sentry isn't able to symbolicate, e.g. off by 1 line numbers, please raise a new issue, thanks.

**Comment by ahmed-ashraf** [comment]
I use this command below to build the app for web
flutter build web --dart-define=environment=dev --web-renderer canvaskit --release --source-maps

Then I use this command to upload source maps
flutter packages pub run sentry_dart_plugin

I got those errors on sentry log
![image](https://github.com/getsentry/sentry-dart/assets/19362719/25583d04-c5f0-4c28-95e6-e4f03c4abb31)

And the stacktrack doesn't show the full code but it views the line of error. How can I fix that?
![image](https://github.com/getsentry/sentry-dart/assets/19362719/285e672c-ab37-4b3f-8eb9-936166ae782e)

**Comment by marandaneto** [comment]
@ahmed-ashraf open a new issue, I'm no longer part of the Sentry staff.

**Comment by JaredEzz** [comment]
I'm using flutter web source maps with no issues in Sentry today - I know this is an old thread but the issues seem to be resolved
